### PR TITLE
board support: Add probing mechanism

### PR DESCRIPTION
**Pull request by doanac**
This introduces a more advanced config file that optionally gives us the ability to automatically probe for the correct board when one isn't specified or found.

Tested on Bubblegum, HiKey, and Dragonboard410c.

### CODE REVIEW

File: lib/include/libsoc_conffile.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * \param char* path - The path to the file
+ * \return NULL on error
+ */
+conffile* conffile_load(const char *path);
+
+/**
+ * \fn conffile *conffile_free(conffile*)
+ * \brief Frees memory allocation in conffile struct
+ */
+conffile* conffile_load(const char *path);
```
Comment (jackmitch) [inline lib/include/libsoc_conffile.h]: double declaration?

File: lib/include/libsoc_conffile.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * \fn conffile *conffile_load(const char *path)
+ * \brief Loads a given configuration file
+ * \param char* path - The path to the file
+ * \return NULL on error
+ */
+conffile* conffile_load(const char *path);
+
+/**
+ * \fn conffile *conffile_free(conffile*)
+ * \brief Frees memory allocation in conffile struct
```
Comment (jackmitch) [inline lib/include/libsoc_conffile.h]: allocated

File: lib/conffile.c [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  conffile *conf = calloc(1, sizeof(conffile));
+  section *stmp = NULL;
+  keyval *ktmp = NULL;
+
+  fp = fopen (path, "r");
+  if (fp)
+    {
+      while(fgets(line, sizeof(line), fp))
+        {
+          if (*line == '#' || *line == '\0' || *line == '\n') continue;
```
Comment (jackmitch) [inline lib/conffile.c]: continue on new line

**Comment by doanac** [comment]
I just did a "push -f" on the branch to fix those issues.

File: lib/Makefile.am [hunk 4]
```diff
@@ -21,4 +21,4 @@ libsoc_la_CPPFLAGS = -I${top_srcdir}/lib/include
 ## interface : source : age
 
 libsoc_la_LDFLAGS = -version-info 5:1:3
-AM_CFLAGS = -DDATA_DIR=\"$(DESTDIR)$(pkgdatadir)\" -DGPIO_CONF=\"@sysconfdir@/libsoc_gpio.conf\"
+AM_CFLAGS = -DDATA_DIR=\"$(DESTDIR)$(pkgdatadir)\" -DGPIO_CONF=\"@sysconfdir@/libsoc.conf\"
```
Comment (jackmitch) [inline lib/Makefile.am]: should this still be called GPIO_CONF?

File: lib/file.c [hunk 5]
```diff
@@ -187,18 +186,31 @@ int file_write_str(char *path, char* buf, int len)
...
 {
+  int fd;
+  struct stat st;
   char *buf;
-  off_t len;
-  int fd = file_open(path, O_RDONLY);
+
+  /* NOTE: lseek(.. SEEK_END) is broken for this special file type on the
+   * bubblegum board, so we have to use stat instead.
+   */
```
Comment (jackmitch) [inline lib/file.c]: if this method works then lets just use this, drop the vendor board specific hack notes as it's not relevant to a mainline geared library where this should be working.

File: contrib/board_files/beaglebone.conf [hunk 6]
```diff
@@ -1,3 +1,6 @@
+[board]
+model = TI AM335x BeagleBone Black
```
Comment (jackmitch) [inline contrib/board_files/beaglebone.conf]: Let me do a little bit of research to see if this is the same for all varients of beaglebone

Comment (jackmitch) [inline contrib/board_files/beaglebone.conf]: Ok, the different varients do have different pin outs to the headers, so this conf file should be named beaglebone_black.conf and it matching on the Black model specifically is correct. We can move it as a seperate patch.

File: lib/Makefile.am (hunk 4 above)
Comment (doanac) [inline lib/Makefile.am]: I changed this to LIBSOC_CONF in latest push

File: lib/file.c (hunk 5 above)
Comment (doanac) [inline lib/file.c]: removed in latest push

File: contrib/board_files/beaglebone.conf (hunk 6 above)
Comment (doanac) [inline contrib/board_files/beaglebone.conf]: I went ahead and fixed it in this push.

**Comment by doanac** [comment]
I think the latest round of comments have all been addressed.

**Comment by jackmitch** [comment]
Couple of cleanup warnings:

make[1]: Entering directory '/mnt/seaRaid/Projects/libsoc.git/lib'
  CC       libsoc_la-gpio.lo
  CC       libsoc_la-spi.lo
  CC       libsoc_la-file.lo
file.c: In function 'file_read_contents':
file.c:193:7: warning: implicit declaration of function 'stat' [-Wimplicit-function-declaration]
   if (stat(path, &st))
       ^
  CC       libsoc_la-i2c.lo
  CC       libsoc_la-pwm.lo
  CC       libsoc_la-board.lo
board.c: In function '_probe_config':
board.c:36:7: warning: implicit declaration of function 'libsoc_debug' [-Wimplicit-function-declaration]
       libsoc_debug(__func__, "No 'model' value found in 'board' section");
       ^
board.c: In function '_probe':
board.c:83:7: warning: implicit declaration of function 'libsoc_warn' [-Wimplicit-function-declaration]
       libsoc_warn("Unable to read directory: %s", confs_dir);
       ^
  CC       libsoc_la-conffile.lo
conffile.c: In function '_create_keyval':
conffile.c:30:7: warning: implicit declaration of function 'libsoc_warn' [-Wimplicit-function-declaration]
       libsoc_warn("Invalid key = value in %s:\n%s\n", path, line);
       ^
  CC       libsoc_la-debug.lo
  CCLD     libsoc.la

**Comment by doanac** [comment]
warning fixed in latest push

File: lib/conffile.c [hunk 7]
```diff
@@ -0,0 +1,176 @@
+#include <ctype.h>
```
Comment (jackmitch) [inline lib/conffile.c]: could we get some general comment descriptions of the functions in here, it took me a while to decode what was actually happening in a couple of them. Just a brief explanation of function and arguments.

File: lib/board.c [hunk 8]
```diff
@@ -1,114 +1,129 @@
...
+		    }
+                }
             }
-          else
-            {
-              cur->next = ptr;
-              cur = cur->next;
-            }
-        }
+        } while (dp != NULL && !bc);
```
Comment (jackmitch) [inline lib/board.c]: can this be

while (dp != NULL && bc == NULL)

**Comment by jackmitch** [comment]
We also need some documentation on the config file format, just a text file in the board_config directory specifying the currently supported sections and what values they are expected to take and defaults etc

File: lib/include/libsoc_conffile.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * \fn conffile *conffile_free(conffile*)
+ * \brief Frees memory allocated in conffile struct
+ */
+void conffile_free(conffile *conf);
+
+/**
+ * \fn const char *conffile_get(conffile *conf, const char *sectname, const char *key, const char *defval)
+ * \brief Returns the value for the given section and key
+ */
```
Comment (jackmitch) [inline lib/include/libsoc_conffile.h]: ah, I see some brief comments here, could they be expanded a bit to say what the arguments are doing, for example this makes no mention of setting a default value if nothing is found.

**Comment by doanac** [comment]
Okay, my latest push:

* while (dp != NULL && bc == NULL) instead of while (dp != NULL && !bc);
* adds a little more documentation to libsoc_conffile.h
* adds new commit with an example.conf: https://github.com/jackmitch/libsoc/pull/47/commits/1df092586250df48ec56ac751ff00df8ba1ed651

File: contrib/board_files/example.conf [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+# Lines starting with "#" are ignored
+
+[board]
+# The dtfile is the file probed to determine if current board matches this
+# board config. Change this value if you system requires a different file for
+# board identification:
+# dtfile = /proc/device-tree/model
+
+# The model value is compared to the contents of the dtfile. If these match,
+# then this board config will be selected.
```
Comment (jackmitch) [inline contrib/board_files/example.conf]: can you state that the _whole_ of the contents of the file have to match, and not just a portion of the file

**Comment by doanac** [comment]
Good suggestion Jack. I've tried to word this better in my latest push (only the HEAD commit has changed)

File: contrib/board_files/example.conf [hunk 11]
```diff
@@ -0,0 +1,21 @@
+# A board config file consists of two required sections, the "board" and "GPIO"
+# Lines starting with "#" are ignored
+
+[board]
+# The dtfile is the file probed to determine if current board matches this
+# board config. Change this value if you system requires a different file for
+# board identification:
+# dtfile = /proc/device-tree/model
+
```
Comment (jackmitch) [inline contrib/board_files/example.conf]: # When attempting an automatic configuration probe at library runtime, the dtfile option contains the path
# to the file which will be read and it's contents compared against the 'model' option. Change this value if you want # your 'model' option to be compared to a different file.

File: contrib/board_files/example.conf [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+# The model value is compared to the contents of the dtfile. If these match,
+# (ie entire_contents_of_dtfile == this value), then this board config will
+# be selected.
```
Comment (jackmitch) [inline contrib/board_files/example.conf]:
```
# The model value is compared to the contents of the file path in 'dtfile'. If the white space trimmed 'model' option,
# is equal to the contents of the whitespace trimmed 'dtfile' then the automatic probe will succed and this
# board config file selected.
```

**Comment by doanac** [comment]
fixed with suggested text.

**Comment by jackmitch** [comment]
Cheers Andy, lets stick it in and see if anything falls out!

**Pull request merged**